activesupport/lib/active_support/tagged_logging.rb

```diff
@@ -29,56 +29,86 @@ module TaggedLogging
     module Formatter # :nodoc:
       # This method is invoked when a log event occurs.
       def call(severity, timestamp, progname, msg)
-        if t = tags_text
-          super(severity, timestamp, progname, t.concat(msg.to_s))
-        else
-          super(severity, timestamp, progname, msg)
-        end
+        super(severity, timestamp, progname, tag_stack.format_message(msg))
       end
 
       def tagged(*tags)
-        new_tags = push_tags(*tags)
+        pushed_count = tag_stack.push_tags(tags).size
         yield self
       ensure
-        pop_tags(new_tags.size)
+        pop_tags(pushed_count)
       end
 
       def push_tags(*tags)
-        tags.flatten!
-        tags.reject!(&:blank?)
-        current_tags.concat tags
-        tags
+        tag_stack.push_tags(tags)
       end
 
-      def pop_tags(size = 1)
-        current_tags.pop size
+      def pop_tags(count = 1)
+        tag_stack.pop_tags(count)
       end
 
       def clear_tags!
-        current_tags.clear
+        tag_stack.clear
       end
 
-      def current_tags
+      def tag_stack
         # We use our object ID here to avoid conflicting with other instances
-        thread_key = @thread_key ||= "activesupport_tagged_logging_tags:#{object_id}"
-        IsolatedExecutionState[thread_key] ||= []
+        @thread_key ||= "activesupport_tagged_logging_tags:#{object_id}"
+        IsolatedExecutionState[@thread_key] ||= TagStack.new
+      end
+
+      def current_tags
+        tag_stack.tags
       end
 
       def tags_text
-        tags = current_tags
-        if tags.one?
-          +"[#{tags[0]}] "
-        elsif tags.any?
-          tags.collect { |tag| "[#{tag}] " }.join
+        tag_stack.format_message("")
+      end
+    end
+
+    class TagStack # :nodoc:
+      attr_reader :tags
+
+      def initialize
+        @tags = []
+        @tags_string = nil
+      end
+
+      def push_tags(tags)
+        @tags_string = nil
+        tags.flatten!
+        tags.reject!(&:blank?)
+        @tags.concat(tags)
+        tags
+      end
+
+      def pop_tags(count)
+        @tags_string = nil
+        @tags.pop(count)
+      end
+
+      def clear
+        @tags_string = nil
+        @tags.clear
+      end
+
+      def format_message(message)
+        if @tags.empty?
+          message
+        elsif @tags.size == 1
+          "[#{@tags[0]}] #{message}"
```

Comment:
There is something odd happening with the 1-tag case.  It is more expensive than it should be, even with this special case:

| version | messages | tags | memsize | objects | strings |
| --- | --- | --- | --- | --- | --- |
| before | 1 | 1 | 562 | 8 | 2 |
| before | 1 | 2 | 550 | 10 | 3 |
| after w/o special case | 1 | 1 | 602 | 9 | 4 |
| after w/o special case | 1 | 2 | 446 | 8 | 4 |
| after w/ special case | 1 | 1 | 522 | 7 | 2 |
| after w/ special case | 1 | 2 | 446 | 8 | 4 |
| before | 3 | 1 | 1.366k | 16 | 2 |
| before | 3 | 2 | 1.330k | 22 | 3 |
| after w/o special case | 3 | 1 | 1.406k | 17 | 4 |
| after w/o special case | 3 | 2 | 938 | 14 | 4 |
| after w/ special case | 3 | 1 | 1.326k | 15 | 2 |
| after w/ special case | 3 | 2 | 938 | 14 | 4 |

So I decided to keep this special case optimization to somewhat offset the underperformance.